pkg/cloud/metadata.go

```diff
@@ -69,13 +75,69 @@ func (m *Metadata) GetOutpostArn() arn.ARN {
 func NewMetadata() (MetadataService, error) {
 	sess := session.Must(session.NewSession(&aws.Config{}))
 	svc := ec2metadata.New(sess)
-	return NewMetadataService(svc)
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return NewMetadataService(svc, clientset)
 }
 
 // NewMetadataService returns a new MetadataServiceImplementation.
-func NewMetadataService(svc EC2Metadata) (MetadataService, error) {
+func NewMetadataService(svc EC2Metadata, clientset kubernetes.Interface) (MetadataService, error) {
 	if !svc.Available() {
-		return nil, fmt.Errorf("EC2 instance metadata is not available")
+		klog.Warningf("EC2 instance metadata is not available")
+		nodeName := os.Getenv("CSI_NODE_NAME")
+		if nodeName == "" {
+			return nil, fmt.Errorf("instance metadata is unavailable and CSI_NODE_NAME env var not set")
+		}
+
+		// get node with k8s API
+		node, err := clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		providerID := node.Spec.ProviderID
+		if providerID == "" {
+			return nil, fmt.Errorf("node providerID empty, cannot parse")
+		}
+
+		awsRegionRegex := "([a-z]{2}(-gov)?)-(central|(north|south)?(east|west)?)-[0-9]"
+		awsAvailabilityZoneRegex := "([a-z]{2}(-gov)?)-(central|(north|south)?(east|west)?)-[0-9][a-z]"
```

Comment:
@ayberk will this owrk in outpost world?

Reply:
i didn't really consider local zones, wavelength, or outposts when i wrote the regular expression. how much does that matter for GA?

Reply:
of those, i think only outpost matters in the sense that we explicitly said we support it. But, on second thought, even if the regex is edited to be more exhaustive to include the outpost case, it will not help them, because outpost users are still reliant on instance metadata being available to retrieve outpost-id..

Reply:
so for now let's just keep it as is I guess

Reply:
cool cool

Reply:
to be specific, we have no way to find out if we are on an outpost and to retrieve outpostArn in case metadata is disabled.

Reply:
What does the `providerId` look like?

There's a case to be made here to allow customers to set outpost-arn themselves so they can use it without the metadata service. I don't think it's a big deal for now but something to consider.

Reply:
`providerId` looks like `aws:///us-west-2b/i-abcdefgh123456789`